utils/src/test/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcherTest.kt

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.lwdataprovider
+
+import com.exactpro.th2.common.grpc.Direction
+import com.exactpro.th2.common.util.toInstant
+import com.exactpro.th2.dataprovider.lw.grpc.DataProviderService
+import com.exactpro.th2.dataprovider.lw.grpc.MessageGroupsSearchRequest
+import com.exactpro.th2.dataprovider.lw.grpc.MessageSearchResponse
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.DEFAULT_REQUEST_INTERVAL
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.with
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.withBook
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.withRequestInterval
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.withSessionAlias
+import com.exactpro.th2.lwdataprovider.MessageSearcher.Companion.withSessionGroup
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+import org.mockito.kotlin.any
+import org.mockito.kotlin.argumentCaptor
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.times
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.verifyNoMoreInteractions
+import org.mockito.kotlin.whenever
+import java.lang.IllegalArgumentException
+import java.time.Duration
+import java.time.Instant
+import java.util.function.Function
+import kotlin.math.ceil
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertNull
+import kotlin.test.assertTrue
+
+class MessageSearcherTest {
+    private val service: DataProviderService = mock { }
+    private val searcher = MessageSearcher(service, REQUEST_INTERVAL)
+
+    @AfterEach
+    fun afterEach() {
+        verifyNoMoreInteractions(service)
+    }
+
+    @Test
+    fun `test iterations`() {
```

Comment:
I believe we should not test the internal staff (like the number of the filter invocations). Instead, we should (I think) configure mock properly (check the request parameters) and check the end result (whether we found the message or not). Also, I believe if we create distinct MessageSearchResponse results it will be better for test - we can identify the exact response (right now every response looks the same)

Reply:
If you see it fits here you are welcome to leave it. I am just saying that instead of testing different cases for overall logic we are testing the particular internal stuff